### PR TITLE
finish new dynamic client and deprecate old dynamic client

### PR DESCRIPTION
**Pull request by deads2k**
Builds on a couple other pulls.  This completes the transition to the new dynamic client.

@kubernetes/sig-api-machinery-pr-reviews 
@caesarxuchao @sttts 

```release-note
The old dynamic client has been replaced by a new one.  The previous dynamic client will exist for one release in `client-go/deprecated-dynamic`.  Switch as soon as possible.
```

### CODE REVIEW

**Comment by deads2k** [comment]
/hold

just holding so we don't merge this without twiddling qps on the GC controller.

**Comment by caesarxuchao** [comment]
Can we leave the old dynamic client interface there for one quarter, and give people a deprecation notice? With a quick search, there are quite a few users: https://github.com/search?p=1&q=%22dynamic.newclient%22&type=Code&utf8=%E2%9C%93

**Comment by caesarxuchao** [comment]
Discussed with deads2k offline, we'll move the old dynamic client to client-go/deprecated-dynamic and keep it for a quarter.

**Comment by deads2k** [comment]
> Discussed with deads2k offline, we'll move the old dynamic client to client-go/deprecated-dynamic and keep it for a quarter.

done, release note added.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ const ResourceResyncTime time.Duration = 0
 // up to date as the notification is sent.
 type GarbageCollector struct {
 	restMapper    resettableRESTMapper
-	dynamicClient dynamic.DynamicInterface
+	dynamicClient dynamic.Interface
```
Comment (sttts) [inline pkg/controller/garbagecollector/garbagecollector.go]: we should have kept the interface name. Would have made this easier. Maybe we can restore it in a follow-up?

File: pkg/kubectl/resource/scheme.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+)
+
+// dynamicCodec is a codec that wraps the standard unstructured codec
+// with special handling for Status objects.
```
Comment (sttts) [inline pkg/kubectl/resource/scheme.go]: is this old code?!

Comment (deads2k) [reply pkg/kubectl/resource/scheme.go]: > is this old code?!

Yeah, moved to point of use.

**Comment by deads2k** [comment]
@sttts split commits as requested.

File: staging/src/k8s.io/client-go/dynamic/fake/simple.go [hunk 3]
```diff
@@ -145,23 +192,65 @@ func (c *dynamicResourceClient) UpdateStatus(obj *unstructured.Unstructured) (*u
 	return ret, err
 }
 
-func (c *dynamicResourceClient) Delete(name string, opts *metav1.DeleteOptions) error {
-	_, err := c.client.Fake.
-		Invokes(testing.NewDeleteAction(c.resource, c.namespace, name), &metav1.Status{Status: "dynamic delete fail"})
+func (c *dynamicResourceClient) Delete(name string, opts *metav1.DeleteOptions, subresources ...string) error {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/dynamic/fake/simple.go]: out of curiosity: do we have subresource deletion anywhere?

Comment (deads2k) [reply staging/src/k8s.io/client-go/dynamic/fake/simple.go]: > out of curiosity: do we have subresource deletion anywhere?

it's logically possible.  nothing immediately comes to mind though.

**Comment by deads2k** [comment]
@sttts @caesarxuchao's concern was addressed and the commits are split for review now.  Review and I'll rebase?

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 1 above)
Comment (sttts) [reply pkg/controller/garbagecollector/garbagecollector.go]: :+1: now with multiple commits this is even visible :)

**Comment by sttts** [comment]
/approve
lgtm

Needs a rebase.

**Comment by deads2k** [comment]
approval is for new client-go package that chao and I agreed on.  We own the sub-piece.  approving.

File: staging/src/k8s.io/client-go/dynamic/dynamic_util.go [hunk 4]
```diff
@@ -1,87 +0,0 @@
...
-
-	"k8s.io/apimachinery/pkg/api/meta"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/runtime/schema"
-)
-
-// NewDiscoveryRESTMapper returns a RESTMapper based on discovery information.
-func NewDiscoveryRESTMapper(resources []*metav1.APIResourceList) (*meta.DefaultRESTMapper, error) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/dynamic/dynamic_util.go]: So `deprecated-dynamic` doesn't have all the interfaces the "old" dynamic client had. I think removing this one is ok, a search on github doesn't show any real use case.

**Review by caesarxuchao — COMMENTED** [review]
lgtm. Sorry I can't get to this earlier.

**Pull request merged**